Separates tx-state and store cursors

tinwelint opened the pull request:
By introducing storage cursors that the implementations of kernel API cursors contain and provide tx-state around. The storage cursors are similar, but they don't share the same interfaces and they also expose logical data as to make it storage agnostic.

Physical details that leaked from the current storage into the storage API (StorageEngine, StorageReader and friends) have been removed. The implementation of the storage cursors also moved to the recordstorage package so that most things can be package private.

fickludd committed:
Break out store cursors from default cursors

tinwelint committed:
Extract interface to StoreNodeCursor

and moves some physical node reading details into it

tinwelint committed:
Extract interface to StorePropertyCursor

and moves some physical node reading details into it

tinwelint committed:
Extract interface to relationship/group store cursors

and simplifies implementations by avoiding lots of complexity
introduced by RelationshipReferenceEncoding.

tinwelint committed:
Moves store cursors to the the recordstorage package

tinwelint committed:
Removal of old storage cursors

And replacing few remaining usages with the new storage cursors.

tinwelint committed:
Removed unused methods from StorageReader

tinwelint committed:
Further removed unused classes and methods

tinwelint committed:
Removed next(filter) from storage cursors

So that the storage cursors can be simplified even further.
The check for removed in tx-state moved into the kernel API cursors,
where they belong.

Also javadoced the storage cursors and created a simple hierarchy
of StorageCursor.

tinwelint committed:
License and checkstyle

fickludd commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/Read.java:
```diff
@@ -265,48 +265,6 @@ long lockingNodeUniqueIndexSeek( IndexReference index, IndexQuery.ExactPredicate
-     * @param cursor
-     *         the cursor to use for consuming the results.
-     */
-    void relationshipGroups( long nodeReference, long reference, RelationshipGroupCursor cursor );
```
Could you please refrain from changing the Kernel API while creating the Store API? We think we have a very good reason for allowing dereferenced relationship and property reads, and we went to lots of extra pains during the last year to make it possible. That functionality needs to be intact.

tinwelint replied:
Oh OK, what is the use case for dereferenced access? It complicated the implementation and wasn't used so I removed it. My idea of it is that allowing dereferenced access exposes more physical details than passing in a cursor to be placed at that point instead. I can bring it back if you really need it. Would just be fun to learn why it would be preferred over passing in a cursor.

tinwelint replied:
I can see that it's now trivial to bring back those methods when the refactoring has gone through, so I'll do that!

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordNodeCursor.java:
```diff
@@ -30,7 +30,7 @@
 import org.neo4j.storageengine.api.StorageNodeCursor;
 
-public class StoreNodeCursor extends NodeRecord implements StorageNodeCursor
+class RecordNodeCursor extends NodeRecord implements StorageNodeCursor
```
Now this is a name that makes complete sense. 👏

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/StoreAbstractRelationshipCursor.java:
```diff
@@ -1,125 +0,0 @@
-/**
- * Base cursor for relationships.
- */
-abstract class StoreAbstractRelationshipCursor implements Cursor<RelationshipItem>, RelationshipItem
```
Getting rid of all these old cursors is an epic win!

fickludd commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/Groups.java:
```diff
@@ -38,6 +38,11 @@
+    public Groups()
+    {
+
+    }
```
Whaa?

tinwelint replied:
Haha, this was a constructor usage trick I pulled for something completely different... I forgot to remove it.

fickludd commented on community/kernel/src/main/java/org/neo4j/storageengine/api/StorageCursor.java:
```diff
@@ -0,0 +1,31 @@
+    /**
+     * Resets the current data for this cursor so that it returns an unused uninitialized item.
+     */
+    void reset();
```
This is probably my fault, by why do we have reset? Shouldn't close be enough?

tinwelint replied:
You may be right, I'll give it a shot.

fickludd reviewed (changes requested):
This time around I think the refactor seems very successful. The vast number of removed classes is incredible. 

My only problem with the PR as it stand is that you have removed the dereferenced accesses of properties and relationships. This is something that we really think we want for the runtime and fought hard to implement. Please bring that back. Was is just for simplification that it was removed or did you discover some fundamental challenge with it?

And finally I would really like to see triggered e.g. LDBC and Kernel API benchmarks on this just to have some more confidence that we are not regressing.

fickludd commented:
Btw, I have reviewed this for architecture only. I have not verified spelling of javadocs or correctness of all migration and new methods. I suggest we merge this as soon as benchmarking and tests agree (and dereferencing is back). Then I and someone else can browse over the new storage API and implementations and generate further clean-up PRs as necessary. This way we can reduce conflict risk.

tinwelint commented:
Exactly, I agree with the idea of merging as soon as possible since getting it 100% right isn't the big point with this PR, it's doing the extraction which is intrusive and we can iterate on it once it's in.

tinwelint committed:
Brings back removed kernel API dereferenced accessor methods

Some of the reference encoding logic is split, so that most of it is
in the kernel code and only select parts remain as store cursor
responsibilities.

tinwelint committed:
Removes unnecessary StorageCursor#reset() method

tinwelint commented:
@fickludd how about now?

fickludd commented:
@tinwelint This looks mighty nice to me, but I can't approve due to the repo-change. If you move it over, I'll approve the new one.
